### DISCUSSION TITLE
Is it possible to to ignore certain characters ("," ".", "-", "")

### DISCUSSION THREAD

**Discussion by Somnium99** [pgroonga/pgroonga#403] [Q&A]
We need to be able to find texts ignoring certain characters like : `.`, `'`, `''`, `"`, `-`.
Example:
1. `Stratford-upon-Avon` = `Stratford upon Avon`
2. `Planetarium No. 1` = `Planetarium No 1`
3. `St. Peter` = `St Peter` (and preferably match also `Saint Peter`)
4. `Forrat's Chocolate` = `Forrat"s Chocolate` = `Forrats Chocolate`

I'm using the `&@` operator and this is our index:

```
CREATE INDEX proonga_title_original_idx ON some_table
    USING pgroonga (title pgroonga_text_full_text_search_ops_v2)
    WITH (tokenizer = 'TokenBigram', normalizer='NormalizerNFKC100')
    WHERE title IS NOT NULL;
```

And it doesn't work. I can only make it work with the `&@*` operator but then the similarity search is too loose.

Is it possible using pgroonga?

**Comment by kou** [comment]
Could you also provide `CREATE TABLE`/`INSERT` for the sample data?

**Comment by Somnium99** [comment]
Sure, here is an example.
(Each insert row representing titles that should be equal in terms of search result)

```
CREATE TABLE places (
	title       text
);

INSERT INTO places (title)
VALUES 
       ('Stratford-upon-Avon'), ('Stratford upon Avon'),
       ('Planetarium No. 1'), ('Planetarium No 1'), 
       ('St. Peter'), ('St Peter'),
       ('Forrat''s Chocolate'), ('Forrat"s Chocolate'), ('Forrats Chocolate'), ('Forrat''''s Chocolate');

CREATE INDEX proonga_places_title_idx ON places
   USING pgroonga (title pgroonga_text_full_text_search_ops_v2)
   WITH (tokenizer = 'TokenBigram', normalizer='NormalizerNFKC100')
   WHERE title IS NOT NULL;

```

**Reply by komainu8** [reply]
Thanks!

How about use "loose_symbol" and "loose_blank" as below?

```sql
CREATE TABLE places (
	title       text
);

INSERT INTO places (title)
VALUES 
       ('Stratford-upon-Avon'), ('Stratford upon Avon'),
       ('Planetarium No. 1'), ('Planetarium No 1'), 
       ('St. Peter'), ('St Peter'),
       ('Forrat''s Chocolate'), ('Forrat"s Chocolate'), ('Forrats Chocolate'), ('Forrat''''s Chocolate');

CREATE INDEX proonga_places_title_idx ON places
   USING pgroonga (title pgroonga_text_full_text_search_ops_v2)
   WITH (tokenizer = '
           TokenNgram("loose_symbol", true, "loose_blank", true)',
	   normalizer='NormalizerNFKC150')
   WHERE title IS NOT NULL;

SELECT * FROM places WHERE title &@ 'Stratford-upon-Avon';
        title        
---------------------
 Stratford-upon-Avon
 Stratford upon Avon
(2 rows)

SELECT * FROM places WHERE title &@ 'Stratford upon Avon';
        title        
---------------------
 Stratford-upon-Avon
 Stratford upon Avon
(2 rows)

SELECT * FROM places WHERE title &@ 'Planetarium No. 1';
       title       
-------------------
 Planetarium No. 1
 Planetarium No 1
(2 rows)

SELECT * FROM places WHERE title &@ 'Planetarium No 1';
       title       
-------------------
 Planetarium No. 1
 Planetarium No 1
(2 rows)

SELECT * FROM places WHERE title &@ 'St. Peter';
   title   
-----------
 St. Peter
 St Peter
(2 rows)

SELECT * FROM places WHERE title &@ 'St Peter';
   title   
-----------
 St. Peter
 St Peter
(2 rows)

SELECT * FROM places WHERE title &@ E'Forrat\'s Chocolate';
        title        
---------------------
 Forrat's Chocolate
 Forrat"s Chocolate
 Forrats Chocolate
 Forrat''s Chocolate
(4 rows)

SELECT * FROM places WHERE title &@ E'Forrat\"s Chocolate';
        title        
---------------------
 Forrat's Chocolate
 Forrat"s Chocolate
 Forrats Chocolate
 Forrat''s Chocolate
(4 rows)

SELECT * FROM places WHERE title &@ 'Forrats Chocolate';
        title        
---------------------
 Forrat's Chocolate
 Forrat"s Chocolate
 Forrats Chocolate
 Forrat''s Chocolate
(4 rows)
```

See also following documents about "loose_symbol" and "loose_blank".

loose_symbol: https://groonga.org/docs/reference/tokenizers/token_ngram.html#loose-symbol
loose_blank: https://groonga.org/docs/reference/tokenizers/token_ngram.html#remove-blank

**Reply by Somnium99** [reply]
It seems to be working exactly like I wanted, thank you!

Just a question regarding the tokenizer :
We used to use the `NormalizerNFKC100` only because otherwise Cyrillic titles didn't work, and that seemed to solve it.
I see now that you used `NormalizerNFKC150`.  
Are there any considerations to be taken here regarding that?
Can we safely use it?

**Reply by kou** [reply]
`NormalizerNFKC*` are based on Unicode. `NormalizerNFKC150` is based on Unicode 15.0. It's the latest supported Unicode. So we use it.

In general, we can safely use newer `NormalizerNFKC*` because existing characters will not changed/removed. (Backward compatible.)
You can check it by yourself because Unicode provides many information including past versions. See also:
* https://unicode.org/versions/
* https://unicode.org/ucd/

**Comment by kou** [comment]
We can use https://pgroonga.github.io/reference/functions/pgroonga-query-expand.html for `St` == `Saint`.

**Reply by Somnium99** [reply]
Is there a way to make `St` = `Saint` be more "global" ?
Because with query-expand, if I understand it correctly, each record has to have the synonyms column populated.

    id     |      title    |   synonyms  
    1      |   St. Peter   |   [Saint]
    2      |   St. Marie   |   [Saint]
    
So if I have multiple records with St., that means each and every record has to have the "Saint" in the synonyms column.

**Reply by kou** [reply]
You just need `St` -> `Saint` mapping. If you want to accept `St.` and `Saint` as query too, you need `St.` -> `Saint` and `Saint` -> `St.` mapping too.

```sql
CREATE TABLE synonyms (
  term text,
  synonyms text[]
);

CREATE INDEX synonyms_term
          ON synonyms
       USING pgroonga (term pgroonga_text_term_search_ops_v2);

INSERT INTO synonyms VALUES ('St', ARRAY['St', 'Saint']);
INSERT INTO synonyms VALUES ('St.', ARRAY['St.', 'Saint']);
INSERT INTO synonyms VALUES ('Saint', ARRAY['Saint', 'St.']);
```

You need to use `&@~` https://pgroonga.github.io/reference/operators/query-v2.html not `&@` because query expansion uses query syntax https://groonga.org/docs/reference/grn_expr/query_syntax.html :

```sql
SELECT * FROM places WHERE title &@~ pgroonga_query_expand('synonyms', 'term', 'synonyms', 'St');
SELECT * FROM places WHERE title &@~ pgroonga_query_expand('synonyms', 'term', 'synonyms', 'St.');
SELECT * FROM places WHERE title &@~ pgroonga_query_expand('synonyms', 'term', 'synonyms', 'Saint');
```
